pkg/controllers/execution/execution_controller.go

```diff
@@ -161,6 +156,26 @@ func (c *Controller) syncWork(ctx context.Context, clusterName string, work *wor
 	return controllerruntime.Result{}, nil
 }
 
+func (c *Controller) handleWorkDelete(ctx context.Context, work *workv1alpha1.Work, cluster *clusterv1alpha1.Cluster) error {
+	if ptr.Deref(work.Spec.PreserveResourcesOnDeletion, false) {
+		klog.V(4).InfoS("Preserving resource on deletion from work(%s/%s) on cluster(%s)", work.Namespace, work.Name, cluster.Name)
+		return nil
+	}
+
+	// Abort deleting workload if cluster is unready when unjoining cluster, otherwise the unjoin process will be failed.
```

Comment:
❤️

Reply:
That's existing code 😆

Reply:
I'm blind, I though you added comments for this mess :D

Reply:
sounds like a good follow-up PR to clean it up